Fixed #32819 -- Established relationship between form fields and their errors.

smithdc1 opened the pull request:
ticket-32819

smithdc1 commented on django/forms/boundfield.py:
```diff
@@ -295,11 +295,17 @@ def build_widget_attrs(self, attrs, widget=None):
-            and self.field.help_text
             and not self.use_fieldset
             and self.auto_id
+            and not self.is_hidden  # Thoughts?
```
I think that we should not add `aria-describedby` when the field is hidden. 

Adding help_text to a hidden input I think we could consider a user error. However, hidden inputs could well have validation errors. I don't think we want to associate those to the input given there's no way for the input to be edited. 

These errors are usually shown at the top of the form instead. I'm not sure how they should be announced but it seems to me like that should be a separate PR / change if needed.

CxrruptedSoul replied:
jj

felixxm replied:
This change looks like a separate cleanup that we could do in advance. As far as I'm aware, it won't make any difference since all `hidden` elements are not visible for both the accessibility tools and browsers presentation layer. We will only reduce the size of the generated HTML.

nessita replied:
I agree that this guard adding should be done as an orthogonal change to this PR.

smithdc1 replied:
Merged in c187f5f9242b681abaa199173e02066997439425

smithdc1 commented on docs/ref/forms/api.txt:
```diff
@@ -1003,29 +1007,42 @@ method you're using:
+        <input type="checkbox" name="cc_myself" id="id_cc_myself" checked>
     </div>
 
+If a custom ``aria-describedby`` is provided when defining the widget this will
+be used in place of the default value. Consideration will need to be given how
+to associate errors with their inputs to allow these descriptions to be read to
+screen reader users.
```
This is rather horrid. Suggestions welcome. 

To help explain my concern -- if you added `aria-describedby` to your `widget.attrs` when defining the form that will be used in place of any thing that django's going to add. That is django won't add a reference to the errors. However, adding an additional `aria-describedby` only when the field has errors it a lot more logic for the user to write. I'm not sure I can explain all of that here, at least not succinctly.

nessita replied:
I'm not sure how we can explain that either, somehow it feels like we should solve this for them. At the very least we should provide a complete example about how to override `aria-describedby` in a correct manner? (I'm honestly not sure what would be the correct way to do this :-))

sarahboyce replied:
As a ugly suggestion but to get us thinking, `Widget` can have something like
```python
    def get_aria_describedby(self, has_errors=False):
        return self.attrs.get("aria-describedby")
```
and we have in `build_widget_attrs`
```python
        widget_aria_describedby = self.widget.attrs.get_aria_describedby(
            has_errors=bool(self.errors)
        )
        if widget_aria_describedby:
            attrs["aria-describedby"] = widget_aria_describedby
```
Then `get_aria_describedby` can be documented and overwritten for custom widgets?

smithdc1 replied:
With the new `aria_describedby` property and the upcoming easing of boundfield customisation I am much less concerned now. There will be a reasonable escape hatch if folk need it.

MHLut replied:
You can put multiple, space-separated IDs in `aria-describedby`. We could append the error ID to the value already provided by the user.

nessita reviewed (commented):
Thank you @smithdc1 for this work, it definitely seems non trivial. I feel that we need to design a plan/have clear docs for when a custom `aria-describedby` is given.

sarahboyce reviewed (commented):
Thank you for this @smithdc1 
I have a suggestion and I'd be interested to see what you think 👍

smithdc1 commented:
@sarahboyce  -- Circling back around to this one. Is it worth moving the first commit to a separate PR and reviewing/merging that first? (probably need some tweaks to the release notes)?

No rush -- hope DjangoCon is going well!

smithdc1 commented on django/forms/boundfield.py:
```diff
@@ -295,12 +295,17 @@ def build_widget_attrs(self, attrs, widget=None):
-            attrs["aria-describedby"] = f"{self.auto_id}_helptext"
+            aria_describedby = []
+            if self.help_text:
+                aria_describedby.append(f"{self.auto_id}_helptext")
```
Both the [UK Gov.uk design service](https://design-system.service.gov.uk/components/error-message/) and [Wagtail](https://github.com/wagtail/wagtail/blob/23275a4cef4d36bb311abe315eb9ddfc43868e8b/wagtail/admin/panels/field_panel.py#L281-L285) have hint text ahead of errors in their examples/code. 

I think _generally_ we can rely on the UK government design service (GDS) as a reference implementation.

I'd therefore lean to going the same was as those two references unless we get a steer from the Accessibility team that contradicts those sources?  What do you think?

sarahboyce replied:
[W3C](https://design-system.w3.org/styles/form-errors.html) has errors above helptext and within the label span but the `aria-describedby` after the help text.
But I think that method means the error is still announced before the help text (as it's within the label area) 
This semi relates to https://github.com/django/django/pull/16975

> I'd therefore lean to going the same was as those two references unless we get a steer from the Accessibility team that contradicts those sources? What do you think?

Yes I agree, there's no references I can find that `aria-describedby` should be before help text :+1:

smithdc1 commented on django/forms/boundfield.py:
```diff
@@ -295,12 +295,17 @@ def build_widget_attrs(self, attrs, widget=None):
         if (
             not attrs.get("aria-describedby")
             and not widget.attrs.get("aria-describedby")
-            and self.field.help_text
             and not self.use_fieldset
             and self.auto_id
             and not self.is_hidden
         ):
-            attrs["aria-describedby"] = f"{self.auto_id}_helptext"
+            aria_describedby = []
+            if self.help_text:
+                aria_describedby.append(f"{self.auto_id}_helptext")
+            if self.errors:
+                aria_describedby.append(f"{self.auto_id}_error")
+            if aria_describedby:
+                attrs["aria-describedby"] = " ".join(aria_describedby)
```
I really like the suggesion of pulling it out into a property. This is much better. 👍 

I think this line here though is a backward incompatible change. See this [comment](https://github.com/django/django/pull/16185#issuecomment-1570975992) and other notes in that PR. The conclusion was that priority would be given to `aria-describedby` defined in userland. If you are setting this then you already know what you're doing and Django adding more will likely make things worse. 

``` diff
-            'aria-describedby="custom-description" id="id_username">'
-            "</div><div>"
+            'aria-describedby="custom-description id_username_helptext" '
+            'id="id_username"></div><div>'
```

smithdc1 commented:
I think this is getting close -- but I need to have another look over the docs. 

Let me know if you want to pull out the change adding the new `aria_describedby` property as a separate commit / PR. The changes there (without adding error references), and the changes to the HTML templates should be able to be merged without any changes to the tests.

sarahboyce commented:
> I think this is getting close -- but I need to have another look over the docs.

Agreed :star: I think we'll get this into 5.2 :crossed_fingers: 

> Let me know if you want to pull out the change adding the new `aria_describedby` property as a separate commit / PR. The changes there (without adding error references), and the changes to the HTML templates should be able to be merged without any changes to the tests.

I think that's another good idea :heart:

sarahboyce commented on django/forms/boundfield.py:
```diff
@@ -289,20 +289,26 @@ def build_widget_attrs(self, attrs, widget=None):
+                attrs["aria-describedby"] = aria_describedby
         return attrs
 
+    @property
```
Maybe we use `@cached_property`  so that `{% if field.aria_describedby %} aria-describedby="{{ field.aria_describedby }}"{% endif %}` builds the string once :thinking:

smithdc1 replied:
Thinking about this today. I'm not so sure -- see the example below the final output doesn't have the error reference. 

I appreciate this is a bit tenuous. 

``` python
In [6]: class ContactForm(forms.Form):
   ...:     name = forms.CharField(help_text="Help Text!")
   ...:

In [7]: f = ContactForm({'name':"David"})

In [8]: print(f)
<div>
    <label for="id_name">Name:</label>
<div class="helptext" id="id_name_helptext">Help Text!</div>
<input type="text" name="name" value="David" required aria-describedby="id_name_helptext" id="id_name">
</div>

In [9]: f.add_error("name", "some additional error")

In [10]: print(f)
<div>
    <label for="id_name">Name:</label>
<div class="helptext" id="id_name_helptext">Help Text!</div>
<ul class="errorlist" id="id_name_error"><li>some additional error</li></ul>
<input type="text" name="name" value="David" required aria-invalid="true" aria-describedby="id_name_helptext" id="id_name">
</div>

```

sarahboyce replied:
Ok, let's have a test that covers this case
I feel like it could be cached in the help text only case and then, we shouldn't cache it for errors which are more dynamic

smithdc1 commented:
Rebased... To do:

- [ ] Review docs
- [ ] Add test as per https://github.com/django/django/pull/17520#discussion_r1883688400
- [ ] Is this now fixed rather than refs?


Anything else?

smithdc1 commented on django/forms/boundfield.py:
```diff
@@ -305,6 +305,8 @@ def aria_describedby(self):
         if self.auto_id and not self.is_hidden:
             if self.help_text:
                 aria_describedby.append(f"{self.auto_id}_helptext")
+            if self.errors:
```
Thank you for helping get the change to just this!

sarahboyce commented:
@django/accessibility could I request that this PR gets a prioritized review? :pray:
I would love to land this in Django 5.2 (feature freeze Jan 15th - 4 weeks today). Based off the screen reader testing I did, it looks good to me but would like to get an approval from the team :heart:

MHLut commented:
Thank you team for this change!

This looks mostly good; I only miss references to non-input form controls, like `<select>` and `<textarea>`.

I would like the tests to cover a broader set of form controls, beyond `<input>`.

smithdc1 commented:
> Is this now fixed rather than refs?

I reviewed the ticket and think this can be called 'fixed'.

> I would like the tests to cover a broader set of form controls, beyond <input>.

Thank you for the feedback.  I've added some additional tests.

knyghty commented:
Reading those form tests is quite tricky, lot of plaintext in here. Wish we had a better way to write these.

I would wait if Marijke thinks anything is missing but it looks great to me and I'll be very happy to have this in 5.2 🚀

sarahboyce commented:
@smithdc1 I added small tweaks (you can [see these here](https://github.com/django/django/compare/ebc45a8af511b84af98d8473f0b0693a12376e8b..6af75bb7c76c71fa6e596738a2c26966f3b15f4c) - shout if there's something you don't like).
I debated about whether we should update the docs for [BoundField.errors](https://docs.djangoproject.com/en/5.1/ref/forms/api/#django.forms.BoundField.errors) as it mentions `aria-invalid`. I think it's not neccessary as the information is captured elsewhere but I am happy for someone else to weigh in an opinion there

smithdc1 commented:
Super. Any comments on docs would be very much welcomed. 

I've now got the same feeling I have at work when I'm checking a presentation of numbers in it. I can spot things the first or second time but after a while I struggle to spot even the most obvious issue. 

I'll need help getting it over the line from here. 🙏

knyghty commented on docs/ref/forms/api.txt:
```diff
@@ -1003,23 +1005,44 @@ method you're using:
+default value.
+
+If the widget is rendered in a ``<fieldset>`` then ``aria-describedby`` is
+added to this element, otherwise it is added to the widget's ``<input>``.
```
I guess it's not necessarily a literal `<input>` element as discussed earlier. I'm not sure if it's worth broadening this or if the intent is clear enough.

sarahboyce replied:
Good point, added a tweak :+1:

MHLut commented:
With the addition of textarea and select in the tests, I'm OK with this!

I only wonder if the mix of British and American English spelling of colo(u)r will be an issue, but that's obviously not a showstopper!

It would be great if this change makes it to `5.2`.

sarahboyce commented:
> I only wonder if the mix of British and American English spelling of colo(u)r will be an issue

Good spot! Updated

sarahboyce reviewed (approved):
Thank you everyone :star:

felixxm commented on docs/ref/forms/api.txt:
```diff
@@ -1003,23 +1005,45 @@ method you're using:
-    <div>Subject:
-      <ul class="errorlist"><li>This field is required.</li></ul>
-      <input type="text" name="subject" maxlength="100" required aria-invalid="true">
+    >>> f = ContactForm(data)
```
```suggestion
    >>> ContactForm(data)
```

sarahboyce replied:
Good spot! I needed to either print it or add `.as_div()`. Gone for `.as_div()`

felixxm commented on docs/ref/forms/api.txt:
```diff
@@ -1003,23 +1005,45 @@ method you're using:
+
+.. versionchanged:: 5.2
+
+    ``aria-describedby`` was added to associate errors with its input.
```
Maybe I'm wrong :thinking: but do we have a mix of singular/plural forms here?
```suggestion
    ``aria-describedby`` was added to an associate error with its input.
```
or
```suggestion
    ``aria-describedby`` was added to associate errors with their input.
```

felixxm replied:
Ah OK, one input many errors, silly me :upside_down_face:
